HHH-19695 invalid SQL generation under DB2 dialect

michaelfranz opened the pull request:
Invalid SQL generated with using DB2 dialect.

The setup:

- An entity declares an attribute using a formula.
- The formula contains SQL.
- The SQL contains a DB2-specific or non-ansi keyword such as `first` or `next`

For example:
```
@Formula(“select count(*) from table where fetch_count > 5 fetch first row only")
private int calc;
```

Under Hibernate 7.0.6 this SQL will be rendered as follows:

`select count(*) from table v1_0 where fetch_count > 5 fetch v1_0.first row only`

The part  `v1_0.first` is not valid SQL and demonstrates that Hibernate has treated `first` as though it were a column reference.

This pull request fixes the issue.

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

<!-- Hibernate GitHub Bot issue links start -->
<!-- THIS SECTION IS AUTOMATICALLY GENERATED, ANY MANUAL CHANGES WILL BE LOST -->
https://hibernate.atlassian.net/browse/HHH-19695
<!-- Hibernate GitHub Bot issue links end -->

gavinking commented:
The problem with this, I imagine (I didn't check), is that `first` and `next` aren't really reserved words on Db2, and can be used as column names.

michaelfranz commented:
> The problem with this, I imagine (I didn't check), is that `first` and `next` aren't really reserved words on Db2, and can be used as column names.

I think you'll find that they are:
https://www.ibm.com/docs/en/db2-for-zos/12.0.0?topic=words-reserved

And to be fair, there are undoubtedly several other reserved words that might come up in the same problem space. If in future people want to add these, this PR makes it clear where in the Hibernate codebase these need to be added.

michaelfranz commented:
> But there are thousands or more likely hundreds of thousands of programs which use Hibernate on Db2 _are) using `first` or `next` as a column name, and this change has the potential to break any of those programs.

I find this observation to be absolutely on the money. So although 7.0.6 breaks _our_ installation because it turns 'fetch first' into 'fetch <alias>.first', the proposed solution may end up breaking other installations, which is obviously not acceptable.

I find your suggestion about solving the problem using recognised bigrams of `fetch first` and `fetch next` to be a better approach to the one I've so far proposed. 

I have updated Template in accordance with your suggestion. I have essentially added logic to detect the start and end of `FETCH FIRST/NEXT`. While full BNF of this clause is `FETCH { FIRST | NEXT } <row count> ROWS ONLY` I have kept (or tried to keep) the implementation minimal and in the spirit of the existing code. Most importantly, I have added an array of  fairly comprehensive tests to `TemplateTest`.

The final observation I would make is that the problem is clearly far more serious that I originally realised. Hibernate 7 breaks ALL applications which use FETCH in the context I outlined in the original ticket. The problem and the solution apply to everyone. I will reformulate HHH-19695 accordingly.

gavinking commented:
> The final observation I would make is that the problem is clearly far more serious that I originally realised. 

Oh OK, the issue description did not mention that this was a regression.

The relevant commit was 29224ed1482910b54767f24a3912ae28ec302661.
